OSSM-8200 Document how to scope the mesh with Discovery Selectors (#84469)
openshift/openshift-docs: modules/ossm-using-discoveryselectors-scope-service-mesh.adoc

// Module included in the following assemblies:
// install/ossm-installing-openshift-service-mesh.adoc

:_mod-docs-content-type: Procedure
[id="ossm-discoveryselectors-scope-service-mesh_{context}"]
= Using DiscoverySelectors to scope a Service Mesh
:context: ossm-discoveryselectors-scope-service-mesh

If you know which namespaces to include in the {SMProductShortName}, configure `discoverySelectors` during or after installation by adding the required selectors to the MeshConfig resource of {istio}. For example, configure {istio} to discover only namespaces labeled `istio-discovery=enabled`.

.Prerequisites
* The {SMProduct} operator is installed
* An {istio} CNI resource is created

.Procedure
. Add a label to the namespace containing the {istio} control plane, for example, the `istio-system` system namespace.
+
[source,terminal]
----
$ oc label namespace istio-system istio-discovery=enabled
----

. Modify the `{istio}` control plane resource to include a `discoverySelectors` section with the same label.
+
[source,yaml]
----

> kind: Istio

Reviewer: It looks like there were some issues copying over this example code... See how the example config was updated [draft doc here](https://github.com/openshift-service-mesh/sail-operator/tree/main/docs/ossm/create-mesh) - procedure step 2. 

The `Istio` CRD could be quite large with a lot of config that's not relevant, so I wasn't sure how much we should show. Looking at other examples, it looks like the common thing is to show the resource headers  (apiVersion, kind, maybe metadata) and then only the relevant configuration that needs to be updated, but I'll look for your guidance here.

This part should definitely not be included though, as it's not relevant to the code being changed and could cause problems if the user overwrites their existing config with these values:
```
updateStrategy:
        type: InPlace
      version: v1.23.0
```

Author: @longmuir Agree with your points. I have updated the example config following the draft doc. For the last point, I have a suggestion. If the user needs to change the version then it can be added as a step (wherever you need it). For example:

. Change the version in the {istio} resource:
+
[source,yaml]
----
kind: Istio
spec:
  version: 1.23.2
  updateStrategy: 
    type: InPlace
 ----
 
If you think that's not needed, I can ignore the comment. Do let me know what you think! :)

// Module included in the following assemblies:
// install/ossm-installing-openshift-service-mesh.adoc

:_mod-docs-content-type: Procedure
[id="ossm-discoveryselectors-scope-service-mesh_{context}"]

> = Using DiscoverySelectors to scope a Service Mesh

Reviewer 2: ```suggestion
= Scoping a Service Mesh by using discovery selectors
```

:context: ossm-discoveryselectors-scope-service-mesh

> If you know which namespaces to include in the {SMProductShortName}, configure `discoverySelectors` during or after installation by adding the required selectors to the MeshConfig resource of {istio}. For example, configure {istio} to discover only namespaces labeled `istio-discovery=enabled`.

Reviewer: ```suggestion
If you know which namespaces to include in the {SMProductShortName}, configure `discoverySelectors` during or after installation by adding the required selectors to the `meshConfig.discoverySelectors` section of the `{istio}` resource. For example, configure {istio} to discover only namespaces labeled `istio-discovery=enabled`.
```
Apologies for only seeing this now - the reference ot "MeshConfig resource..." looks like an upstream reference, so it would be more accurate to describe the configuration value changed in the `Istio` resources which is part of the product (as shown in step 2).

Author: Made this change. Thanks!

.Prerequisites

> * The {SMProduct} operator is installed

Reviewer 2: ```suggestion
* The {SMProduct} operator is installed.
```

> * An {istio} CNI resource is created

Reviewer 2: ```suggestion
* An {istio} CNI resource is created.
```

.Procedure
. Add a label to the namespace containing the {istio} control plane, for example, the `istio-system` system namespace.
+
[source,terminal]
----
$ oc label namespace istio-system istio-discovery=enabled
----

. Modify the `{istio}` control plane resource to include a `discoverySelectors` section with the same label.
+
[source,yaml]
----
kind: Istio
apiVersion: sailoperator.io/v1alpha1
metadata:
  name: default
spec:
  namespace: istio-system
  values:
    meshConfig:
      discoverySelectors:
        - matchLabels:
            istio-discovery: enabled
----

. Apply the Istio CR:
+
[source,terminal]
----
$ oc apply -f istio.yaml
----

> . Ensure that all namespaces that will contain workloads that are to be part of the {SMProductShortName} have both the `discoverySelector` label and, if desired, the appropriate {istio} injection label.

Reviewer 2: ```suggestion
. Ensure that all namespaces that will contain workloads that are to be part of the {SMProductShortName} have both the `discoverySelector` label and, if needed, the appropriate {istio} injection label.
```
See https://www.ibm.com/docs/en/ibm-style?topic=word-usage#desire

> 

Reviewer 2: Should the following note be nested under step 4?

Author: I don't think so. Th note in the end is a note for all the whole procedure. Also the last step is connected to the **Next Steps** mentioned after this module.

[NOTE]
====

> `discoverySelectors` help restrict the scope of a single {SMProductShortName} and are essential for limiting the control plane scope when deploying multiple {istio} control planes in a single cluster.

Reviewer 2: ```suggestion
Discovery selectors help restrict the scope of a single {SMProductShortName} and are essential for limiting the control plane scope when you deploy multiple {istio} control planes in a single cluster.
```
